### PR TITLE
AssetProcessor: allow to disable initial scan

### PR DESCRIPTION
**Pull request by CyaNinix**
Since the asset processors loads (and thus lock) every gems, we can't work on gems and compile them by leaving it open. However, everytime the asset processor starts, it runs a fast scan to check if any file has changed, which on big projects can take a few minutes, and game/editor won't start until it finished.

When working and testing pure C++ changes, this can waste a lot of time, waiting until the asset processor understands no file has changed.

We added a checkbox to (temporary) disable startup scan, and a button to perform the fast scan (similarly to full scan):
image

### CODE REVIEW

File: Code/Tools/AssetProcessor/native/AssetManager/assetProcessorManager.cpp
Comment (amzn-mike) [inline Code/Tools/AssetProcessor/native/AssetManager/assetProcessorManager.cpp:2892]: The coding standard calls for brackets for all control statements
```suggestion
            for (const AssetFileInfo& fileInfo : filePaths)
            {
                AddKnownFoldersRecursivelyForFile(fileInfo.m_filePath, fileInfo.m_scanFolder->ScanPath());
            }
```

File: Code/Tools/AssetProcessor/native/AssetManager/assetProcessorManager.cpp
Comment (amzn-mike) [inline Code/Tools/AssetProcessor/native/AssetManager/assetProcessorManager.cpp:2896]: I see you added a fast scan button.  Will that button function if this data is cleared?  I didn't see anything to re-populate it.

**Comment by AMZN-stankowi** [comment]
I'd like the UX sig to review this before I dig too deep into the code.

While I agree that conceptually, something like this would be really helpful for rapid iteration in the conditions you've described, my concern is that this is the sort of setting that can cause other problems. If someone has forgotten they've enabled this, they may end up with stale assets and run into problems. It might not be obvious to check if a setting like this is enabled.

I think this feature is useful, it just needs some UX attention to address those concerns, so I'd like the UX sig to be pulled into assist.

File: Code/Tools/AssetProcessor/native/AssetManager/assetProcessorManager.cpp
Comment (CyaNinix) [reply Code/Tools/AssetProcessor/native/AssetManager/assetProcessorManager.cpp:2898]: Yes, those maps are actually only needed while a scan is performed and gets cleared once the scan finishes.

We tested the fast scan button without any issue.

**Comment by CyaNinix** [comment]
> f someone has forgotten they've enabled this, they may end up with stale assets and run into problems. It might not be obvious to check if a setting like this is enabled.

I do agree, maybe I should add a warning on the homepage of the asset processor "warning: startup scan is disabled, changes made while the asset processor will **not** be processed".

**Comment by galibzon** [comment]
UX should jump in, because an alternative approach would be to defer the scan. The Editor would start immediately and the AP would do the scan, albeit with deferred results. This way there wouldn't be any issues with "remembering" to re-active the AP.

**Comment by rainbj** [comment]
Hi Team, so a few bits of feedback on this ticket. Conversations about this topic are welcome if you disagree.

1. We usually only use checkboxes when the answer can be more than one at a time. Example: I like, strawberry, chocolate, AND vanilla ice cream. We do toggle switch for yes or no questions or on and off states. Example: Would you like to turn on the ice cream security camera (on/off) = toggle.
2. The buttons and the toggle switch should follow the same visual pattern. So users don't have to look around to understand / see mixed patterns (currently buttons on bottom of text but checkboxes on left. ) My suggestion is we keep the same pattern as component cards. Titles on left,  actions on right.
3. Updates the sub text of disable startup scan. We should avoid using brackets and just write out the sentence as plainly as possible. I have made a suggestion, but feel free to get a second opinion. 
4. All text should be aligned on the left. Currently, there is a bunch of staggering, and it makes it harder to scan reading. (line color: #616161 )
Asset Processor settings
Asset Processor settings2
5. I have concerns about users forgetting this setting is on and then reporting that the editor is no longer processing assets. 
My suggestion has two ways of resolving this issue. Please do at least 2a) 
     1a Maybe after a system restart we change this flag back to the default value and or some other automated way to revert this change over time.
      2a. We need to visually let the user know inside the editor that this change has taken place. This way, users cannot just forget about this change. (color - #FA2727)
Asset tagging
6. My assumption is that this setting that cannot be persisted / checked in for other users? I would not want other users on the same game team to end up inheriting this setting from another contributor on their team.
7. Field order: I'm unable to make an assumption if these settings are in the correct order. For now, I think it's ok, but we might change it in the future.

**Review by rainbj — CHANGES_REQUESTED** [review]
Please see previous comments

**Comment by CyaNinix** [comment]
> 6. My assumption is that this setting that cannot be persisted / checked in for other users? I would not want other users on the same game team to end up inheriting this setting from another contributor on their team.

This setting is stored at the same place as the fast scan setting, which is local to the computer.

>    2a. We need to visually let the user know inside the editor that this change has taken place. This way, users cannot just forget about this change. (color - #FA2727)

Note that this changes does not disable the asset processor, it will update if any change happens while it's running.

> because an alternative approach would be to defer the scan. The Editor would start immediately and the AP would do the scan, albeit with deferred results. This way there wouldn't be any issues with "remembering" to re-active the AP.

This seems like a much better idea than this PR, however I'm not sure how it should be implemented.

**Comment by lmbr-pip** [comment]
[CyaNinix](https://github.com/CyaNinix) first huge apologies for this getting stalled out. We are all still learning how to run an open source project.

So, mode makes sense to everyone and the only friction point is the UX. Here is a delivery plan we thrashed out:

B1. To deliver as is but remove the "Start Faster Scan" option, so we just bring the "Disable Fast Scan" mode option in. Want to avoid user confusion and users can always restart AP with the option off to perform this action.

With that, we can take this PR. After its in we will then:

A1. Adjust the text based on feedback from UX team and move checkboxes to "toggle" options.
A2. Update the user guide on o3de.org
A3. Add an automation test for the mode.

Please let me know if you want to make change B1 and update PR.

Otherwise SIG/content will find someone to take over this PR to update it to latest dev and bring it across the line. We will ensure you get credited in next release notes regardless.

Feel free to reachout to me on Discord (am SeaPip on O3DE discord) if you want to chat directly.

Thanks again for the contribution and my apologies for the delay.

**Comment by CyaNinix** [comment]
Thank you for the feedback @lmbr-pip 

I removed the fast scan button and resolved conflicts

File: Code/Tools/AssetProcessor/native/utilities/ApplicationManagerBase.cpp
Comment (AMZN-stankowi) [inline Code/Tools/AssetProcessor/native/utilities/ApplicationManagerBase.cpp:310]: You might want to remove this for now while there is no UI for it. Not a big deal to keep it besides any code maintenance gotchas, if nothing calls this for a long time, the logic here becomes stale, and someone starts using it in the future, it could have problems.

**Review by AMZN-stankowi — APPROVED** [review]

**Review by lmbr-pip — APPROVED** [review]

**Comment by lmbr-pip** [comment]
@CyaNinix - Looks like there are some DCO issue to fix here, see https://github.com/o3de/o3de/pull/10056/checks?check_run_id=10030870047
 
 I have triggered the automated review for this PR. Takes about 2hours to build and run the approval testing.

**Comment by lmbr-pip** [comment]
@CyaNinix - Branch failed in AR because branch is a little stale (always the fun of AR, as we test this specific branch but often including parts of O3DE. Is it possible to refresh branch with latest dev, which should fix this particular failure on the Android build?

```
D:/workspace/o3de-extras/Gems/OpenXRVk/Code/Source/OpenXRVkInstance.cpp(362,44): error C2039: 'm_context': is not a member of 'AZ::Vulkan::XRInstanceDescriptor::<unnamed-type-m_outputData>' 
```

https://jenkins.build.o3de.org/blue/organizations/jenkins/O3DE/detail/PR-10056/1/pipeline

**Comment by lmbr-pip** [comment]
Just started AR for this: https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-10056/2/ 

Fingers crossed

**Comment by forkercat** [comment]
> Just started AR for this: https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-10056/2/
> Fingers crossed

I think @CyaNinix you may need to do a merge development to pick up some changes, otherwise, the AR will fail.

**Comment by CyaNinix** [comment]
I just did.

**Comment by lmbr-pip** [comment]
Now failing on Multiplayer tests, which are unconnected to changes here. Have pinged maintainers on that.

Trying AR again (#4)

**Comment by lmbr-pip** [comment]
@CyaNinix thanks for the update, I am setting up AR run and hopefully this time we pass.

**Comment by lmbr-pip** [comment]
Looking to see if @rainbj will remove the "requested change" blocked or if we can get an uber maintainer to clear flag. Sorry should have caught that earlier, but now blocked until that can be removed.

**Pull request merged**
